site/_posts/2021-11-23-announcing-fusionauth-1-31.md

```diff
@@ -0,0 +1,49 @@
+author: Dan Moore
+image: blogs/release-1-31/product-update-fusionauth-1-31.png
+category: blog
+tags: topic-troubleshooting feature-advanced-threat-detection
+excerpt_separator: "<!--more-->"
+---
+
+We're excited to announce the release of version 1.31.0 of FusionAuth. This version shipped on November 18, 2021. 1.31.0 includes a ton of bug fixes and improvements to the Identity Providers feature.
+
+<!--more-->
+
+This release contained a number of features, enhancements, and bug fixes. Please see the [release notes](/docs/v1/tech/release-notes/#version-1-31-0) for a full breakdown of the changes between 1.30 and 1.31. 
+
+There are a few items worth calling out.
+
+## Modify the username and email in a reconcile lambda
+
+FusionAuth lambdas are JavaScript functions that run at specific points in the authentication or authorization process. Identity Providers are external sources of account data, such as Facebook, Google, SAML or OIDC servers. Identity Provider reconcile lambdas run when an Identity Provider returns after a succesful login. 
+
+Previous to this release, with a few exceptions, the username and email claims could not be modified in a reconcile lambda. This increases flexibility when you have a remote identity datasource which does not provide either an email or a username. Sometimes identity linking can suffice, but there are other times when it does not. However, any time code can modify login Ids, make sure you threat model out the ramifications, particularly the danger of inadvertent account takeover.
+
+
+## Access to the id_token in the OIDC reconcile lambda
+
+The OIDC process returns not a single token, but two. The Access Token and the Id Token. Previously, the OIDC reconcile lambda only provided the Access Token. With this release, the JavaScript code running in your lambda now has access to the Id Token. As with the Access Token, the lambda can examine the properties of the Id Token and modify the user or registration based on those properties. An example of data available in the Id Token is Azure AD group membership.
+
+If there is no Id Token available, or if it is signed with an asymmetric key, it will not be available. Your lambda code should be able to handle either case.
+ 
+## The rest of it
+
+Some of the other 20+ enhancements and fixes included in this release:
+
+* Performance improvements were made in the 1.30 point releases. In testing, FusionAuth handled over 2,000 registrations per second on properly sized hardware. These improvements are included in this release.
+* A subtle bug in the SAML implementation, relating to the casing of URL encoded characters in a query string, was squashed.
+* Often you'll want to send users directly to an Identity Provider, skipping the FusionAuth login screen. With this release, you can now do that, using the `idp_hint` parameter, for the Twitter and Apple Identity Providers. They join the Google, Facebook, OIDC and SAML Identity Providers in supporting this parameter.
```

Comment:
I believe that all IDPs support idp_hint now except for HYPR and SAMLv2idpinitiaited now. That matches what we have in the docs here as well https://fusionauth.io/docs/v1/tech/identity-providers/#hints. I think Apple and Twitter were the only outliers from what I can tell in the code.